crates/sui-indexer/src/models/events.rs

```diff
@@ -77,96 +88,138 @@ pub fn read_last_event(pg_pool_conn: &mut PgPoolConnection) -> Result<Option<Eve
+        let mut last_event_envelope = event_envelopes.pop().unwrap();
+        last_event_envelope.next_cursor = Some(next_cursor_val);
+        event_envelopes.push(last_event_envelope);
+        next_cursor = next_cursor_val;
+    } else {
+        // unwrap here are safe because we already checked the length of new_events
+        let next_cursor_event_envelope = event_envelopes.pop().unwrap();
+        let mut last_event_envelope = event_envelopes.pop().unwrap();
+        last_event_envelope.next_cursor = Some(next_cursor_event_envelope.transaction_digest);
+        event_envelopes.push(last_event_envelope);
+        next_cursor = next_cursor_event_envelope.transaction_digest;
+    }
+    let new_events: Vec<NewEvent> = event_envelopes
+        .into_iter()
+        .flat_map(|e| event_envelope_to_events(pg_pool_conn, e))
+        .collect();
+
+    // NOTE: Postgres can take at most 65535 parameters in a single query, one single txn on private testnet
+    // can have thousands of events, thus we will need to batch write here.
+    let event_batch_commit_result = commit_events_impl(pg_pool_conn, new_events);
+    Ok(Some((event_batch_commit_result, next_cursor)))
+}
+
+fn sui_event_to_new_event(
+    sui_event: SuiEvent,
+    sequence_number: i64,
+    transaction_digest: TransactionDigest,
+    timestamp_opt: Option<u64>,
+    next_cursor: Option<TransactionDigest>,
+) -> Result<NewEvent, IndexerError> {
+    let event_json = serde_json::to_string(&sui_event).map_err(|err| {
         IndexerError::InsertableParsingError(format!(
             "Failed converting event to JSON with error: {:?}",
             err
         ))
     })?;
-    let timestamp = NaiveDateTime::from_timestamp_millis(e.timestamp as i64).ok_or_else(|| {
-        IndexerError::DateTimeParsingError(format!(
-            "Cannot convert timestamp {:?} to NaiveDateTime",
-            e.timestamp
-        ))
-    })?;
-
+    let mut naive_date_time_opt = None;
+    if let Some(timestamp) = timestamp_opt {
+        let naive_date_time =
+            NaiveDateTime::from_timestamp_millis(timestamp as i64).ok_or_else(|| {
+                IndexerError::DateTimeParsingError(format!(
+                    "Cannot convert timestamp {:?} to NaiveDateTime",
+                    timestamp
+                ))
+            })?;
+        naive_date_time_opt = Some(naive_date_time);
+    }
     Ok(NewEvent {
-        transaction_digest: format!("{:?}", e.tx_digest),
-        event_sequence: e.id.event_seq,
-        event_time: Some(timestamp),
-        event_type: e.event.get_event_type(),
+        transaction_digest: transaction_digest.base58_encode(),
+        event_sequence: sequence_number,
+        event_time: naive_date_time_opt,
+        event_type: sui_event.get_event_type(),
         event_content: event_json,
-        next_cursor_transaction_digest: None,
-        next_cursor_event_sequence: None,
+        next_cursor_transaction_digest: next_cursor.map(|d| d.base58_encode()),
```

Comment:
Why do we need this on every event row?

Reply:
related to the problem above, the ideal solution would be we can use last transaction digest as an "excluded cursor", otherwise, either we will need a column like this, or a separate log table.

the rationale of the merging can be found in this [PR](https://github.com/MystenLabs/sui/pull/7524)